commit f948b6ecc783526516b5a13825ff047169de32b8
Author:     Yan Cheng <58191769+yanchengnv@users.noreply.github.com>
AuthorDate: Thu Apr 18 13:10:42 2024 -0400
Commit:     GitHub <noreply@github.com>
CommitDate: Thu Apr 18 13:10:42 2024 -0400

    Improve dead client handling (#2506)
    
    * dev
    
    * test dead client cmd
    
    * added more info for dead client tracing
    
    * remove unused imports
    
    * fix unit test
    
    * fix test case
    
    * address PR comments
    
    ---------
    
    Co-authored-by: Sean Yang <seany314@gmail.com>
---
 nvflare/private/fed/server/server_engine.py | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index cc0c3a04..60f28bb7 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -577,13 +577,26 @@ class ServerEngine(ServerEngineInternalSpec):
             data = {"execution_error": execution_error}
             job_id = fl_ctx.get_job_id()
             request = new_cell_message({CellMessageHeaderKeys.JOB_ID: job_id}, data)
-            return_data = self.server.cell.fire_and_forget(
+            self.server.cell.fire_and_forget(
                 targets=FQCN.ROOT_SERVER,
                 channel=CellChannel.SERVER_PARENT_LISTENER,
                 topic=ServerCommandNames.UPDATE_RUN_STATUS,
                 message=request,
             )
 
+    def notify_dead_job(self, job_id: str, client_name: str, reason: str):
+        shareable = Shareable()
+        shareable.set_header(ServerCommandKey.FL_CLIENT, client_name)
+        shareable.set_header(ServerCommandKey.REASON, reason)
+        self.send_command_to_child_runner_process(
+            job_id=job_id,
+            command_name=ServerCommandNames.HANDLE_DEAD_JOB,
+            command_data=shareable,
+            timeout=0.0,
+            optional=True,
+        )
+        self.logger.warning(f"notified SJ of dead-job: {job_id=}; {client_name=}; {reason=}")
+
     def send_command_to_child_runner_process(
         self, job_id: str, command_name: str, command_data, timeout=5.0, optional=False
     ):
@@ -595,7 +608,7 @@ class ServerEngine(ServerEngineInternalSpec):
                     targets=fqcn,
                     channel=CellChannel.SERVER_COMMAND,
                     topic=command_name,
-                    request=request,
+                    message=request,
                     optional=optional,
                 )
                 return None
